app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt

```diff
@@ -94,14 +93,22 @@ class DefaultPagedHistoryProvider(
             } else {
                 history = historyStorage
                     .getVisitsPaginated(
-                        offset,
-                        numberOfItems,
+                        offset.toLong(),
+                        numberOfItems.toLong(),
                         excludeTypes = excludedVisitTypes
                     )
-                    .mapIndexed(transformVisitInfoToHistoryItem(offset.toInt()))
+                    .map { transformVisitInfoToHistoryItem(it) }
             }
 
-            onComplete(history)
+            val positionedHistory = history.mapIndexed { index, item ->
+                when(item) {
+                    is History.Group -> item.copy(position = offset + index + item.items.size)
+                    is History.Metadata -> item.copy(position = offset + index)
+                    is History.Regular -> item.copy(position = offset + index)
+                }
+            }
```

Comment:
Any way we add a check for the positions in our existing tests?

Reply:
Oh, yeah, I should have marked this as a WIP. Still not 100% of the approach. I _think_ this is what we want, and I _think_ this really improves a few cases (including the bug), but need to test/think a bit more.

Reply:
Found a bug in the logic while re-reading the PR - this needs to be converted to a for-loop. E.g. if first item is a group of 10 items, and second item is a regular history item, current behavior will assign position=2 (or 1..? not sure if it's 0-indexed) to the second item, while it _should_ assign 11.

Not a huge deal, just something to fix before landing & cover w/ tests.